docs/source/daml2ts/index.rst

```diff
@@ -0,0 +1,290 @@
+
+The DAML type ``Pair a b`` has been modeled as a TypeScript discriminated union (with a single case). In general, all DAML variants are translated so as we will see in the next section.
+
+Variants
+~~~~~~~~
+
+This is a DAML type for a language of additive expressions.
+
+.. code-block:: daml
+   :linenos:
+
+   data Expr a =
+       Lit a
+     | Var Text
+     | Add (Expr a, Expr a)
+
+In TypeScript, it is represented as a discriminated union.
+
+.. code-block:: typescript
+   :linenos:
+
+   type Expr<a> =
+     |  { tag: 'Lit'; value: a }
+     |  { tag: 'Var'; value: string }
+     |  { tag: 'Add'; value: {_1: Expr<a>, _2: Expr<a>} }
+
+Sum-of-products
+~~~~~~~~~~~~~~~~
+
+Let's slightly modify the ``Expr a`` type of the last section into the following.
+
+.. code-block:: daml
+   :linenos:
+
+   data Expr a =
+       Lit a
+     | Var Text
+     | Add {lhs: Expr a, rhs: Expr a}
+
+Compared to the earlier definition, the ``Add`` case is now in terms of a record with fields ``lhs`` and ``rhs``. This renders in TypeScript like so.
+
+.. code-block:: typescript
+   :linenos:
+
+   type Expr<a> =
+     |  { tag: 'Lit2'; value: a }
+     |  { tag: 'Var2'; value: string }
+     |  { tag: 'Add'; value: Expr.Add<a> }
+
+   namespace Expr {
+     type Add<a> = {
+       lhs: Expr<a>;
+       rhs: Expr<a>;
+     }
+   }
+
+The thing to note is how the definition of the ``Add`` case has given rise to a record type definition ``Expr.Add``.
+
+Enums
+~~~~~
+
+DAML enumerations map naturally to TypeScript.
+
+.. code-block:: daml
+   :linenos:
+
+   data Color = Red | Blue | Yellow
+
+The companion TypeScript type is the following.
+
+.. code-block:: typescript
+   :linenos:
+
+   enum Color {
+     Red = 'Red',
+     Blue = 'Blue',
+     Yellow = 'Yellow',
+   }
+
+Templates and choices
+~~~~~~~~~~~~~~~~~~~~~
+
+Here is a DAML template of a basic 'IOU' contract.
+
+.. code-block:: daml
+   :linenos:
+
+   template Iou
+     with
+       issuer: Party
+       owner: Party
+       currency: Text
+       amount: Decimal
+     where
+       signatory issuer
+       choice Transfer: ContractId Iou
+         with
+           newOwner: Party
+         controller owner
+         do
+           create this with owner = newOwner
+
+``daml2ts`` generates types for each of the choices defined on the template as well as the template itself.
+
+.. code-block:: typescript
+   :linenos:
+
+   type Transfer = {
+     newOwner: daml.Party;
+   }
+
+   type Iou = {
+     issuer: daml.Party;
+     owner: daml.Party;
+     currency: string;
+     amount: daml.Numeric;
+   }
+
+Each template and each template choice also results in the generation of a companion object corresponding to the choice or template type. Here, is a schematic of the ones generated from the ``Iou`` template.
+
+.. code-block:: typescript
+   :linenos:
+
+   const Transfer: daml.Serializable<Transfer> = ({
+     /* ... */
+     }),
+   })
+
+   const Iou: daml.Template<Iou, undefined> & {
```

Comment:
We should very briefly mention that the `undefined` captures the fact that `Iou` has no contract key.

Reply:
✔️